scylla/tests/integration/query_result.rs

```diff
@@ -0,0 +1,297 @@
+
+    // Wrong value type
+    match session.query_iter(query, (1,)).await {
+        Ok(_) => panic!("the query should have failed due to a serialization error"),
+        Err(e) => assert_matches!(e, PagerExecutionError::SerializationError(_)),
+    }
+}
+
+#[tokio::test]
+async fn execute_single_page_should_only_iterate_over_rows_in_current_page() {
+    let (session, ks) = initialize_cluster_two_partitions().await;
+
+    let mut prepared_query = session
+        .prepare(format!("SELECT * FROM {}.t where k0 = ?", ks))
+        .await
+        .unwrap();
+    prepared_query.set_page_size(PAGE_SIZE);
+
+    let paging_state = PagingState::start();
+    let (rs_manual, paging_state_response) = session
+        .execute_single_page(&prepared_query, (PARTITION_KEY1,), paging_state)
+        .await
+        .unwrap();
+    let page_results = rs_manual
+        .into_rows_result()
+        .unwrap()
+        .rows::<(String, i32, i32)>()
+        .unwrap()
+        .collect::<Result<Vec<_>, _>>()
+        .unwrap();
+
+    assert!(page_results.len() <= PAGE_SIZE as usize);
+    match paging_state_response {
+        PagingStateResponse::HasMorePages { state: _ } => {}
+        PagingStateResponse::NoMorePages => {
+            panic!("Expected more pages");
+        }
+    }
+}
+
+#[tokio::test]
+async fn execute_iter_should_iterate_over_all_pages_asynchronously_single_partition() {
+    let (session, ks) = initialize_cluster_two_partitions().await;
+
+    let mut prepared_query = session
+        .prepare(format!("SELECT * FROM {}.t where k0 = ?", ks))
+        .await
+        .unwrap();
+    prepared_query.set_page_size(PAGE_SIZE);
+
+    let query_result = session
+        .execute_iter(prepared_query, (PARTITION_KEY1,))
+        .await
+        .unwrap();
+    let mut iter = query_result.rows_stream::<(String, i32, i32)>().unwrap();
+
+    let mut i = 0;
+
+    while let Ok(Some((a, b, c))) = iter.try_next().await {
+        assert_eq!(a, PARTITION_KEY1);
+        assert_eq!(b, i);
+        assert_eq!(c, i);
+        i += 1;
+    }
+    assert_eq!(i, ROWS_PER_PARTITION);
+}
+
+#[tokio::test]
+async fn execute_iter_should_iterate_over_all_pages_asynchronously_cross_partition() {
+    let (session, ks) = initialize_cluster_two_partitions().await;
+
+    let mut prepared_query = session
+        .prepare(format!("SELECT * FROM {}.t", ks))
+        .await
+        .unwrap();
+    prepared_query.set_page_size(PAGE_SIZE);
+
+    let query_result = session.execute_iter(prepared_query, ()).await.unwrap();
+    let mut iter = query_result.rows_stream::<(String, i32, i32)>().unwrap();
+
+    let mut i = 0;
+    while let Ok(Some((a, b, c))) = iter.try_next().await {
+        if i < ROWS_PER_PARTITION {
+            assert_eq!(a, PARTITION_KEY1);
+        } else {
+            assert_eq!(a, PARTITION_KEY2);
+        }
+        assert_eq!(b, i);
+        assert_eq!(c, i);
+        i += 1;
+    }
```

Comment:
:wrench: This does not ensure that all rows are deserialized correctly.
```suggestion
    while let Some((a, b, c)) = iter.try_next().await.unwrap() {
        if i < ROWS_PER_PARTITION {
            assert_eq!(a, PARTITION_KEY1);
        } else {
            assert_eq!(a, PARTITION_KEY2);
        }
        assert_eq!(b, i);
        assert_eq!(c, i);
        i += 1;
    }
```